src/Wlan.cpp

```diff
@@ -10,173 +10,481 @@
+			handleWifiStateConnectLast();
+			return;
+		case WIFI_STATE_SCAN_CONN:
+			handleWifiStateScanConnect();
+			return;
+		case WIFI_STATE_CONN_SUCCESS:
+			handleWifiStateConnectionSuccess();
+			return;
+		case WIFI_STATE_CONNECTED:
+			handleWifiStateConnected();
+			return;
+		case WIFI_STATE_DISCONNECTED:
+			wifiState = WIFI_STATE_INIT;
+			return;
+		case WIFI_STATE_CONN_FAILED:
+			handleWifiStateConnectionFailed();
+			return;
+		case WIFI_STATE_AP:
+			handleWifiStateAP();
+			return;
+		case WIFI_STATE_END:
+			WiFi.disconnect(true, true);
+			WiFi.mode(WIFI_OFF);
+			return;
+	}
+}
+
+bool Wlan_SetHostname(String newHostname) {
+	// hostname should just be applied after reboot
+	gPrefsSettings.putString("Hostname", newHostname);
+	return true;
+}
+
+char nvs_key[9];
+char* getSSIDKey(uint8_t index) { 
+	snprintf(nvs_key, 9, "SSID_%d", index);
+	return nvs_key;
+}
+char* getPWDKey(uint8_t index) { 
+	snprintf(nvs_key, 9, "WPWD_%d", index);
+	return nvs_key;
+}
+
+bool Wlan_AddNetworkSettings(String ssid, String pwd) {
+	uint8_t len = Wlan_NumSavedNetworks();
+
+	for (uint8_t i=0; i<len; i++) {
+		String issid = gPrefsSettings.getString(getSSIDKey(i));
+
+		if (ssid.equals(issid)) {
+			Log_Printf(LOGLEVEL_NOTICE, wifiUpdateNetwork, ssid.c_str());
+			gPrefsSettings.putString(getPWDKey(i), pwd);
+			return true;
 		}
 	}
 
-	if (Wlan_IsConnected()) {
-		if (millis() - lastPrintRssiTimestamp >= 60000) {
-			lastPrintRssiTimestamp = millis();
-			Log_Printf(LOGLEVEL_DEBUG, "RSSI: %d dBm", Wlan_GetRssi());
+	if (len > 9) {
+		Log_Println((char *) FPSTR(wifiAddTooManyNetworks), LOGLEVEL_ERROR);
+		return false;
+	}
+
+	Log_Printf(LOGLEVEL_NOTICE, wifiAddNetwork, ssid.c_str());
+
+	gPrefsSettings.putString(getSSIDKey(len), ssid);
+	gPrefsSettings.putString(getPWDKey(len), pwd);
+
+	numSavedNetworks++;
+	gPrefsSettings.putUChar("NUM_NETWORKS", numSavedNetworks);
+
+	return true;
+}
+
+
+uint8_t Wlan_NumSavedNetworks() {
+	return numSavedNetworks;
+}
+
+// write ssids from nvs into the passed array. 
+uint8_t Wlan_GetSSIDs(String* ssids) {	
+
+	uint8_t len = Wlan_NumSavedNetworks();
+	if (len > 10) {
+		Log_Printf(LOGLEVEL_ERROR, wifiTooManyNetworks, len, 10);
+		len = 10;
+	}
+	
+	uint8_t i;
+	for (i=0; i<len; i++) {
+		String ssid = gPrefsSettings.getString(getSSIDKey(i));
+		Log_Printf(LOGLEVEL_NOTICE, wifiNetworkLoaded, i, ssid.c_str());
+
+		ssids[i] = ssid;
+	}
+	return i;
+}
+
+String Wlan_GetPwd(String ssid) {
+	uint8_t len = Wlan_NumSavedNetworks();
+
+	for (uint8_t i=0; i<len; i++) {
+		String issid = gPrefsSettings.getString(getSSIDKey(i));
+
+		if(ssid.equals(issid)) {
+			return gPrefsSettings.getString(getPWDKey(i));
 		}
 	}
+	return String();
+}
+
+String Wlan_GetCurrentSSID() {
+	return WiFi.SSID();
+}
+
+String Wlan_GetHostname() {
+	return gPrefsSettings.getString("Hostname", "");
+}
 
-	if (accessPointStarted && dnsServer) {
-		dnsServer->processNextRequest();
+bool Wlan_DeleteNetwork(String ssid) {
```

Comment:
You're right. I'm not sure what happened here, but the count should obviously only be decreased on success.

Right now it does not matter that only the NVS is updated because the variable is seldomly used, but I agree it's bad design.

Reply:
fixed in 1d3c9f6

I'm not a fan of the index based erase. I thought of it initally, but IMO a stable key is better.